step/step.go

```diff
@@ -42,21 +67,227 @@ func New(logger log.Logger, inputParser stepconf.InputParser, commandFactory com
+
+	return saver.Save(cache.SaveCacheInput{
+		StepId:      "save-cache",
+		Verbose:     input.Verbose,
+		Key:         fmt.Sprintf(dockerCacheKeyTemplate, imageName),
+		Paths:       []string{dockerCacheFolder},
+		IsKeyUnique: false,
+	})
+}
+
+func (step DockerBuildPushStep) dockerBuild(input Input) error {
+	step.logger.Infof("Building docker image...")
+
+	if err := step.createCacheFolder(dockerCacheFolder); err != nil {
+		return fmt.Errorf("create cache folder: %w", err)
+	}
+	if err := step.createCacheFolder(dockerCacheFolderTemporary); err != nil {
+		return fmt.Errorf("create cache folder: %w", err)
+	}
+
+	buildkitContainer, err := step.initializeBuildkit(input)
+	if err != nil {
+		return fmt.Errorf("initialize buildkit: %w", err)
+	}
+	defer func() {
+		if err := step.destroyContainer(buildkitContainer); err != nil {
+			step.logger.Errorf("destroy buildx instance: %s", err)
+		}
+	}()
+
+	if err := step.build(input); err != nil {
+		return fmt.Errorf("build docker image: %w", err)
+	}
+
+	// To make sure the cache is not growing indefinitely,
+	// we remove the original cache folder and move the new one to its place
+	if err := step.moveCacheFolder(dockerCacheFolderTemporary, dockerCacheFolder); err != nil {
+		return fmt.Errorf("move cache folder: %w", err)
+	}
+
+	return nil
+}
+
+func (step DockerBuildPushStep) destroyContainer(container string) error {
+	args := []string{
+		"buildx", "rm", "--force", container,
+	}
+	cmd := step.commandFactory.Create("docker", args, nil)
+	out, err := cmd.RunAndReturnTrimmedCombinedOutput()
+	if err != nil {
+		return fmt.Errorf("remove buildx instance %s: %w", out, err)
+	}
+	return nil
+}
+
+func (step DockerBuildPushStep) build(input Input) error {
+	stdout := NewLoggerWriter(step.logger)
+	defer stdout.Flush()
+
+	args := []string{
+		"buildx",
+		"build",
+	}
+
+	if input.BuildArg != "" {
+		for _, arg := range strings.Split(input.BuildArg, "\n") {
+			args = append(args, "--build-arg", arg)
+		}
+	}
+
+	switch {
+	case input.UseBitriseCache:
+		args = append(args, fmt.Sprintf("--cache-from=type=local,src=%s", dockerCacheFolder))
+		args = append(args, fmt.Sprintf("--cache-to=type=local,dest=%s,mode=max,compression=zstd", dockerCacheFolderTemporary))
+	case input.CacheFrom != "":
+		for _, cacheFrom := range strings.Split(input.CacheFrom, "\n") {
+			args = append(args, fmt.Sprintf("--cache-from=%s", cacheFrom))
+		}
+		fallthrough
+	case input.CacheTo != "":
+		for _, cacheTo := range strings.Split(input.CacheTo, "\n") {
+			args = append(args, fmt.Sprintf("--cache-to=%s", cacheTo))
+		}
+	}
+
+	if input.ExtraOptions != "" {
+		for _, option := range strings.Split(input.ExtraOptions, "\n") {
+			// This regex splits the string by spaces, but keeps quoted strings together
+			// Example --build-arg "-X main.version=1.0.0" will be split into --build-arg and "-X main.version=1.0.0"
+			r := regexp.MustCompile(`[^\s"']+|"([^"]*)"|'([^']*)`)
```

Comment:
I'm not against (all) regex, but sounds like this stdlib package could do something like what we need: https://pkg.go.dev/flag